Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/CqlTranslator.java

```diff
@@ -470,9 +521,42 @@ public CqlErrorListener(LibraryBuilder builder, boolean detailedErrors) {
             this.detailedErrors = detailedErrors;
         }
 
+        private VersionedIdentifier extractLibraryIdentifier(cqlParser parser) {
+            RuleContext context = parser.getContext();
+            while (!(context instanceof cqlParser.LibraryContext)) {
+                context = context.parent;
+                if (context == null) {
+                    break;
+                }
+            }
+
+            if (context instanceof cqlParser.LibraryContext) {
+                for (ParseTree pt : ((cqlParser.LibraryContext)context).children) {
+                    if (pt instanceof cqlParser.LibraryDefinitionContext) {
+                        cqlParser.LibraryDefinitionContext ldc = (cqlParser.LibraryDefinitionContext)pt;
+                        if (ldc.qualifiedIdentifier() != null && ldc.qualifiedIdentifier().identifier() != null) {
+                            return new VersionedIdentifier().withId(StringEscapeUtils.unescapeCql(ldc.qualifiedIdentifier().identifier().getText()));
+                        }
+                    }
+                }
```

Comment:
I don't know why I'm being so nit-picky about this section of code, but doesn't Java have a way to find an item in a list based on criteria (as opposed to a big loop)?  I haven't ever used it, but based on Google, I *think* you could do something like:
```java
cqlParser.LibraryDefinitionContext ldc = ((cqlParser.LibraryContext)context).children
    .stream().filter(pt -> pt instanceof cqlParser.LibraryDefinitionContext).findAny();
if (ldc.qualifiedIdentifier() != null && ldc.qualifiedIdentifier().identifier() != null) {
    return new VersionedIdentifier().withId(StringEscapeUtils.unescapeCql(ldc.qualifiedIdentifier().identifier().getText()));
}
```
I don't know.  You might need to add a cast.  Maybe it's not any better.  I just don't like using `for` loops to search for things. ;-)

Reply:
I'm also assuming a `LibraryContext` can only have one `LibraryDefinitionContext`, right?  If so, and if you keep the `for` loop, then you could add a `return null` or a `break` after line 538 above because there's no sense in continuing to loop over the child contexts if you already found the library definition.

Reply:
Agreed, and you're right, there is only one libraryDefinitionContext, so I don't need the looping at all. Fixed, thank you for the feedback!

Reply:
Ha.  Yes, even better!